src/utils/crypto_utils/src/crypto_lib.c

```diff
@@ -441,3 +458,139 @@ void CryptoUtils_FreeCryptoKeyHandle(CryptoKeyHandle key)
 {
     EVP_PKEY_free(CryptoKeyHandleToEVP_PKEY(key));
 }
+
+/**
+ * @brief Computes a SHA256 hash from bytes.
+ *
+ * @param buf The handle to the input bytes to be hashed.
+ * @return CONSTBUFFER_HANDLE The digest buffer handle. Caller will own it and need to call CONSTBUFFER_DecRef() to free.
+ */
+CONSTBUFFER_HANDLE CryptoUtils_CreateSha256Hash(const CONSTBUFFER_HANDLE buf)
+{
+    const size_t Sha256DigestNumBytes = 32;
+    unsigned char hashBuf[Sha256DigestNumBytes];
+    memset(hashBuf, 0, sizeof(hashBuf));
+    const CONSTBUFFER* constbuf = CONSTBUFFER_GetContent(buf);
+
+    CONSTBUFFER_HANDLE out_handle = NULL;
+
+    unsigned char* output_digest = NULL;
+    const EVP_MD* msg_digest = NULL;
+    EVP_MD_CTX* mdctx;
+
+    if (buf == NULL)
+    {
+        return NULL;
+    }
+
+    // Initialize OpenSSL digest context
+    if ((mdctx = EVP_MD_CTX_new()) == NULL)
+    {
+        goto done;
+    }
+
+    msg_digest = GetEvpMdFromShaAlg("sha256");
+
+    // Initialize OpenSSL digest operation
+    if (1 != EVP_DigestInit_ex(mdctx, msg_digest, NULL))
+    {
+        goto done;
+    }
+
+    // Hash the input data
+    if (1 != EVP_DigestUpdate(mdctx, constbuf->buffer, constbuf->size))
+    {
+        goto done;
+    }
+
+    // Finalize the digest and get the output
+    if (1 != EVP_DigestFinal_ex(mdctx, &hashBuf[0], NULL))
+    {
+        goto done;
+    }
+
+    output_digest = (unsigned char*)malloc(Sha256DigestNumBytes);
+    if (output_digest != NULL)
+    {
+        memcpy(output_digest, &hashBuf[0], Sha256DigestNumBytes);
+
+        out_handle = CONSTBUFFER_Create(output_digest, Sha256DigestNumBytes);
```

Comment:
isn't there an API to alloc and assign a const buffer handle so that you dont need to separately do malloc/memcpy, or am i thinking of somethign else?

Reply:
It already does the copy into extra bytes allocated in the const buffer. I was being silly as it's unnecessary malloc+memcpy. Fixed in upstream with: https://github.com/Azure/iot-hub-device-update/pull/489/commits/d66ff0eb8c767556d9b4cb95ed00fefe97d017e6